nni/algorithms/compression/pytorch/pruning/iterative_pruner.py

```diff
@@ -0,0 +1,574 @@
+            Optional('op_types'): [str],
+            Optional('op_names'): [str]
+        }], model, logger)
+
+        schema.validate(config_list)
+
+    def calc_mask(self, wrapper, wrapper_idx=None):
+        """
+        Calculate the mask of given layer.
+        Scale factors with the smallest absolute value in the BN layer are masked.
+        Parameters
+        ----------
+        wrapper : Module
+            the layer to instrument the compression operation
+        wrapper_idx: int
+            index of this wrapper in pruner's all wrappers
+        Returns
+        -------
+        dict | None
+            Dictionary for storing masks, keys of the dict:
+            'weight_mask':  weight mask tensor
+            'bias_mask': bias mask tensor (optional)
+        """
+
+        config = wrapper.config
+        freq = config.get('frequency', 1)
+
+        if wrapper.if_calculated:
+            return None
+
+        if not self.now_epoch % freq == 0:
+            return None
+
+        target_sparsity = self.compute_target_sparsity(config)
+        new_mask = self.masker.calc_mask(sparsity=target_sparsity, wrapper=wrapper, wrapper_idx=wrapper_idx)
+        
+        if new_mask is not None:
+            wrapper.if_calculated = True
+
+        return new_mask
+
+    def compute_target_sparsity(self, config):
+        """
+        Calculate the sparsity for pruning
+        Parameters
+        ----------
+        config : dict
+            Layer's pruning config
+        Returns
+        -------
+        float
+            Target sparsity to be pruned
+        """
+
+        initial_sparsity = 0
+        end_epoch = self.training_epochs
+        freq = config.get('frequency', 1)
+        self.target_sparsity = final_sparsity = config.get('sparsity', 0)
+        
+        if initial_sparsity >= final_sparsity:
+            logger.warning('your initial_sparsity >= final_sparsity')
+            return final_sparsity
+
+        if end_epoch == 1 or end_epoch <= self.now_epoch:
+            return final_sparsity
+            
+        span = ((end_epoch - 1) // freq) * freq
+        assert span > 0
+        self.target_sparsity = (final_sparsity +
+                           (initial_sparsity - final_sparsity) *
+                           (1.0 - ((self.now_epoch) / span)) ** 3)
+        return self.target_sparsity
+
+    def update_epoch(self, epoch):
+        """
+        Update epoch
+        Parameters
+        ----------
+        epoch : int
+            current training epoch
+        """
+
+        if epoch > 0:
+            self.now_epoch = epoch
+            for wrapper in self.get_modules_wrapper():
+                wrapper.if_calculated = False
+
+    def compress(self):
+        for epoch in range(self.training_epochs):
+            self.update_epoch(epoch)
+            self._trainer(self.bound_model, optimizer=self.optimizer, criterion=self._criterion, epoch=epoch)
+            self.update_mask()
```

Comment:
Is current position of function `update_mask` appropriate? I think put it before trainer is more reasonable since training is the process of finetuning after masking weight. If it was put after training, we will skip the last iteration finetuning and the results of training may not align with testing.

Reply:
This is mainly because some pruners need to get some information during training, like `TaylorFOWeightFilterPruner`, so we `update_mask` after `train`.

And maybe we need to discuss if users should finetune the pruned model by themselves.